datalad/local/foreach.py

```diff
@@ -0,0 +1,275 @@
+)
+from datalad.interface.common_opts import (
+    jobs_opt,
+    recursion_flag,
+    recursion_limit,
+)
+from datalad.interface.results import get_status_dict
+from datalad.interface.utils import eval_results
+from datalad.support.constraints import (
+    EnsureBool,
+    EnsureNone,
+)
+from datalad.support.parallel import (
+    ProducerConsumer,
+    ProducerConsumerProgressLog,
+    no_parentds_in_futures,
+    no_subds_in_futures,
+)
+from datalad.support.param import Parameter
+from datalad.utils import (
+    SequenceFormatter,
+    getpwd,
+    shortened_repr,
+)
+
+lgr = logging.getLogger('datalad.local.foreach')
+
+
+@build_doc
+class ForEach(Interface):
+    r"""Run a command on the dataset and/or each of its sub-datasets.
+
+    Each dataset's repo runner is reused to execute the commands, so they
+    are are executed in the top directory of a corresponding dataset.
+    In contrast, Python commands (expressions) are evaluated without changing directory.
+    TODO: unify via explicit --chdir? but not sure if worth it
+
+    WARNING: Python expressions are `eval`ed within the scope of this process
+    and provided existing objects.
+
+    """
+    # TODO:     _examples_ = [], # see e.g. run
+
+    _params_ = dict(
+        cmd=Parameter(
+            args=("cmd",),
+            nargs=REMAINDER,
+            metavar='COMMAND',
+            doc="""command for execution. A leading '--' can be used to
+            disambiguate this command from the preceding options to DataLad."""),
+        python=Parameter(
+            args=("--python",),
+            action="store_true",
+            doc="""if given, must be a boolean flag indicating whether
+            the `cmd` is Python expressions to be evaluated, instead of
+            an external command to be executed"""),
+        # Following options are taken from subdatasets
+        dataset=Parameter(
+            args=("-d", "--dataset"),
+            doc="""specify the dataset to operate on.  If
+            no dataset is given, an attempt is made to identify the dataset
+            based on the input and/or the current working directory""",
+            constraints=EnsureDataset() | EnsureNone()),
+        # TODO: assume True since in 99% of use cases we just want to operate
+        #  on present subdatasets?  but having it explicit could be good to
+        #  "not miss any".  But `--fulfilled true` is getting on my nerves
+        fulfilled=Parameter(
+            args=("--fulfilled",),
+            doc="""if given, must be a boolean flag indicating whether
+            to report either only locally present or absent datasets.
+            By default subdatasets are reported regardless of their
+            status""",
+            constraints=EnsureBool() | EnsureNone()),
+        recursive=recursion_flag,
+        recursion_limit=recursion_limit,
+        # does without any outputs handling
+        # Not sure yet if worth supporting here
+        # contains=Parameter(
```

Comment:
It seems like it would be useful, but of course it's easy to tack on later if you're not sure.

Reply:
yeah, I might as well add it, thank you!

Reply:
For large datasets (think UKbiobank), this could make all the difference in terms of performance.

Reply:
ok, "enabling" the --contains option

Reply:
wasdone